### DISCUSSION TITLE
Advanced packageRules logic

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#21684] [Suggest an Idea]
**What would you like Renovate to be able to do?**

Add tag-based semantics, for more flexible metadata and rules.

**Did you already have any implementation ideas?**

1. Tag dependencies/updates, e.g. `baseBranch:master`, `manager:poetry`, `depType:dependencies`, `depType:dependencies`, `datasource:pypi`, `updateType:patch`, `breaking:true`, `lockfile-only:true`.

2. Add `matchAnyTags`, `matchAllTags`, and `excludeTags` fields to `packageRules`

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

Existing packageRules solve most of these problems.

Also:
 - Is it a problem that we'll be *duplicating* some capabilities?
 - Will we one day want more complex logic for tags, and if so then should we add it now already? e.g. nested AND OR NOT logic borrowed from SQL syntax
 - What about matches we support in packageRules already with "partial" matching - paths, sourceUrlPrefixes, package patterns?

**Is this a feature you'd be interested in implementing yourself?**

Yes, but I'd like feedback

**Comment by rarkins** [comment]
A subset of SQL WHERE syntax could work nicely, especially if we extend it to support regex.

`package = 'angular' OR package MATCHREGEX '^@angular/'`

`manager = 'poetry' OR manager = 'pipenv' AND (package MATCHREGEX 'foo' OR package MATCHREGEX 'bar')`

`depType = 'dependencies' AND updateType = 'breaking' AND packageFile MATCHGLOB 'backend/*' AND sourceUrl MATCHPREFIX 'https://github.com/foo/'`

Perhaps use `EQUALS` instead of `=` for consistency? e.g.

`package EQUALS 'angular' OR package MATCHREGEX '^@angular/'`

Actually the above is no longer really tag-based :)

**Comment by rarkins** [comment]
Postgresql syntax: https://www.postgresql.org/docs/12/functions-matching.html#FUNCTIONS-POSIX-REGEXP

I can see two approaches:

1. Keep it same as SQL wherever possible, or
2. Make it most understandable for Renovate users

I'm leaning towards (2)

**Comment by viceice** [comment]
we can use [antlr](https://www.antlr.org) for generating a parser for our custom sql language.

**Comment by rarkins** [comment]
Another thing that would be nice to support is being able to query against `depName` or `lookupName`. Today `packageNames` and `packagePatterns` query against `depName` only.

**Comment by rarkins** [comment]
Instead of an SQL-like approach, here is an alternative syntax: https://github.com/aws/event-ruler/

Java only, but could potentially be ported.
